Version2 reformat obs workflow

valeriupredoi opened the pull request:
OK so we have a quasi working version of the reformat workflow. Why working? Beacuse now you can run `esmvaltool -cr config-reformat-file.yml dummy_recipe.yml` and esmvaltool will go in reformat mode rather than recipe/diagnostic mode (dummy_recipe is just an empty file) - it will not do anything recipe-y but instead it will call `reformat.cmor_reformat` with options set in the config-reformat file. This, in turn, will set the I/O parameters and look for the reformat_obs_XXX script (currently only ncl works, but we can add a whole bunch more languages, if supported by the tool), and further, actually executes `ncl /path/to/reformat_obs_XXX.ncl`. And here is why it doesn't work - I don't know how to pass the ncl script the I/O parameters, mind you, they are all set correctly in `reformat.cmor_reformat()` where we are actually calling the ncl script - @mattiarighi can you maybe make the ncl scripts read a params file that `reformat` can write for them (instead of v1's interface_data/ncl.interface), or set ncl command line options? Both have readily available correct values in the python `reformat` wrapper ready to be passed to ncl

mattiarighi commented:
Can you remove all reformat scripts and just create a dummy one for testing?
I would prefer to move them over as long as I port them.

A dummy NCL script could just contain
```
begin

  print(input_path)
  print(output_path)

end
```

mattiarighi commented on esmvaltool/config-reformat-file.yml:
```diff
@@ -10,6 +10,6 @@ input_dir: ./TEST_INPUT
-datasets: ['ESRL']
+datasets: ['DUMMY']
 # Variables
 variables: ['pr']
```
There no need to specify the variable here, this is done within each script.

valeriupredoi replied:
ok cool, will remove

valeriupredoi replied:
actually var could be useful since we don't want to repeat the reformatting each time we add a new variable file?

mattiarighi replied:
Yes, but then we will end up defining a configuration file for each dataset, or the user would need to edit the configuration file each time a reformat script is run. This also mean he would have to dig through the dataset and check which variable are needed.

If you set the variable (or variable list) directly in the reformat script the user does not need to care about anything (which is what he usually likes). You have to see the reformat scripts as black boxes: I download the data, I put them somewhere, I defined this somewhere as RAWOBS inputdir, I run the script. Done.

mattiarighi commented on esmvaltool/config-reformat-file.yml:
```diff
@@ -10,6 +10,6 @@ input_dir: ./TEST_INPUT
 # Destination directory
 output_dir: ./TEST_OUTPUT
 # Dataset
-datasets: ['ESRL']
```
Can we pass this as input argument instead?

mattiarighi commented on esmvaltool/config-reformat-file.yml:
```diff
@@ -10,6 +10,6 @@ input_dir: ./TEST_INPUT
 # Destination directory
 output_dir: ./TEST_OUTPUT
```
Can we use the original `config-file.yml` for this?

mattiarighi replied:
What about using `rootpath` as a dedicated input-dir, just adding an extra project RAWOBS?
```
rootpath:
  CMIP5: [~/cmip5_inputpath1, ~/cmip5_inputpath2]
  OBS: ~/obs_inputpath
  RAWOBS: ~/rawobs_inputpath
  default: ~/default_inputpath
```
And then we can add another command line flag to run the tool in reformat mode, as you suggest (I would prefer to call this cmorize mode, though).

mattiarighi commented:
Given my comments above, maybe we can get rid of `config-reformat-file.yml` and just use `config-user.yml` as in the standard execution.

Ideally, we could also pass the dataset name as extra argument without need for a dummy recipe
```
esmvaltool -c config-user.yml DUMMY
```
or something similar along the lines of this @jvegasbsc [old suggestion](https://github.com/ESMValGroup/ESMValTool/issues/232#issuecomment-406574796).

Does it sound reasonable?

valeriupredoi commented:
`esmvaltool -cr config-reformat-file.yml DUMMY` works fine; even `esmvaltool -cr config-reformat-file.yml MISS_OCTOBER` works :grin: 

@mattiarighi I can make config-user work instead of config-reformat no problemo, but are we sure we don't want any more options specific for reformat? we don't want to add a whole lot of reformat-specific stuff in config-user. Personally, I'd keep things well separate

mattiarighi commented:
In my opinion, having all input directories defined in the same location is more consistent.
Together with the output path (also defined in config-user.yml), this is the only setting which is shared among all reformat scripts. All the rest is dataset specific and is defined internally in each script.

valeriupredoi commented:
okidokey will add RAWOBS to config-user and rid of config-reformat; also change -cr to -cm (--cmorize)

valeriupredoi commented:
OK @mattiarighi changes made, removed the config-reformat file, added a RAWOBS option in config-file, now you can run the cmorization mode with `esmvaltool -cm -c config-user.yml DUMMY` - I still need to pass the python-constructed I/O information to the ncl script, but all else works a-okay :grin:

mattiarighi commented on esmvaltool/reformat_scripts/constants.ncl:
```diff
@@ -0,0 +1,881 @@
+;;#############################################################################
```
This file can be removed.

mattiarighi commented on esmvaltool/reformat_scripts/recognized_units.dat:
```diff
@@ -0,0 +1,69 @@
+#########################################################################
```
This file can be removed.

mattiarighi commented on esmvaltool/reformat_scripts/recognized_vars.dat:
```diff
@@ -0,0 +1,260 @@
+###############################################################################
```
This file can be removed.

mattiarighi commented on esmvaltool/reformat_scripts/reformat_obs_DUMMY.ncl:
```diff
@@ -0,0 +1,9 @@
+;; this is a dummy test script
```
I would put the reformat scripts in `esmvaltool/cmor/cmorize_obs/` and name them as `cmorize_obs_[dataset].ncl`.

mattiarighi commented on esmvaltool/_main.py:
```diff
@@ -129,21 +119,80 @@ def main(args):
+    timestamp_format = "%Y-%m-%d %H:%M:%S"
+
+    logger.info(
+        "Starting the ESMValTool Reformat Mode v%s at time: %s UTC",
```
`Reformat Mode` --> `in CMORization mode`

mattiarighi commented on esmvaltool/reformat.py:
```diff
@@ -0,0 +1,45 @@
+
+def cmor_reformat(config, logvar):
+    """Run the oldskool v1 reformat scripts."""
+    logvar.info("Running the oldschool reformatting scripts.")
```
`Running the CMORization scripts for the observations"
These will be actually v2 once ported.

mattiarighi commented on esmvaltool/reformat.py:
```diff
@@ -0,0 +1,45 @@
+    # assemble i/o information
+    project_info = {}
+    raw_obs = config["rootpath"]["RAWOBS"][0]
+    for _, datasets, _ in os.walk(raw_obs, followlinks=True):
```
We should not loop over all datasets in RAWOBS, but just call the one specified as command line argument (e.g., DUMMY).

valeriupredoi replied:
@mattiarighi there is no such command line option as DUMMY, DUMMY replaces recipe, I thought the whole point was to batch reformat a whole lot of datasets in RAWOBS

mattiarighi replied:
See [my comment above](https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-430669906). The idea is to pass the dataset name as an argument. So if I pass "DUMMY", the tool should call `cmorize_obs_DUMMY`.

In any case, we do not want to process all datasets in RAWOBS.

valeriupredoi replied:
ok cool, will get that done, wouldn't this be crazy since if you want to cmorize 20 models you will have to run 20 times? anyway, we can probably make it read the datasets to cmorize from a list

mattiarighi replied:
Yes, but going through all datasets in RAWOBS is even crazier.
A list would be the compromise solution. :handshake:

mattiarighi commented:
Thanks @valeriupredoi, I think this is the way to go.
See some comments above.

valeriupredoi commented:
hey @mattiarighi now the framework talks to ncl and is fully functional (for the DUMMY case, it prints out the i/o vars); I will now address your comments above and prettify the code

mattiarighi commented:
We need support for all languages used for the scripts in [v1](https://github.com/ESMValGroup/ESMValTool/tree/development/reformat_scripts/obs). This means NCL, python, bash and csh.

However, I'm not really keen to `bash` and `csh` scripts, since they are calling `cdo` and `nco`, which are not part of our installation. Maybe we can skip those for the moment and I can try to translate the scripts written in these languages to NCL.

Long story short, we need python.

valeriupredoi commented:
okies, I will integrate Python, after lunch :grin: for now try `esmvaltool -cm -c config-user.yml DUMMY` and be awed by the beauty of two ncl variables printed to screen :rofl:

mattiarighi commented:
It is still looping over all directories `<dir>` in my RAWOBS path and trying to run:

`/esmvaltool/cmor/cmorize_obs/cmorize_obs_<dir>.ncl`

valeriupredoi commented:
OK @mattiarighi I fixed a bug that you may have already noticed, restructured the code and now it can run with Python reformatters as well, I also tested with multiple dummy datasets and the script calling and data finding is done correctly - I think you are ready to fire up those cmorization scripts, man!

valeriupredoi commented:
well @mattiarighi - I give you Rome :rofl: the last commit gives you the possiblity to run on a list of datasets like `esmvaltool -cr -c config-user.yml DUMMY -ol DATA1,DATA2...` (this works with `-ol DATA1` only as well) or without the `-ol` flag, it will run across all the RAWOBS dir; I reckon the way to run all over a certain RAWPBS dir is useful because a certain user might want to cmorize say 10 datasets in a directory and doesn't want to write all 10 of them in the command line. What say you? Oh BTW DUMMY refers to a bogus recipe name, it has nothing to do with the datasets, I just don't know how to make the code not ask for a recipe even if it doesn't even exist physically

mattiarighi commented:
I'm a bit confused.
I've tried to run: 
```
esmvaltool -cm -c config-user.yml DUMMY -ol DUMMY
```
I understand the the first `DUMMY` is a workaround to replace the recipe, but I would expect the second `DUMMY` to run the script `cmor/cmorize_obs/cmorize_obs_DUMMY.ncl`.
This is what is shown on stdout:
```
2018-10-19 07:30:11,789 UTC [36187] INFO    Starting the ESMValTool CMORization Mode v2.0a1 at time: 2018-10-19 07:30:11 UTC
2018-10-19 07:30:11,789 UTC [36187] INFO    ----------------------------------------------------------------------
2018-10-19 07:30:11,789 UTC [36187] INFO    INPUT DIR    = /ESMVal/obs/RAW
2018-10-19 07:30:11,790 UTC [36187] INFO    OUTPUT DIR = /ESMValTool/v2_output/CMORization_20181019_073011
2018-10-19 07:30:11,790 UTC [36187] INFO    ----------------------------------------------------------------------
2018-10-19 07:30:11,790 UTC [36187] INFO    Running the CMORization scripts.
2018-10-19 07:30:11,791 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
2018-10-19 07:30:11,792 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
2018-10-19 07:30:11,792 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
2018-10-19 07:30:11,793 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
2018-10-19 07:30:11,793 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
2018-10-19 07:30:11,793 UTC [36187] INFO    Ending the Earth System Model Evaluation Tool v2.0a1 at time: 2018-10-19 07:30:11 UTC
2018-10-19 07:30:11,793 UTC [36187] INFO    Time for running the recipe was: 0:00:00.004039
2018-10-19 07:30:11,794 UTC [36187] INFO    Finished the reformatting module
2018-10-19 07:30:11,794 UTC [36187] INFO    Run was succesful
```

What am I doing wrong?

valeriupredoi commented:
That's because you dont have any data directory called DUMMY in RAWOBS,
Mattia. The -ol is the list of datasets that need to be cmorized but if
there is no dataset with those names in RAWOBS, it will not run

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 19 Oct 2018, 08:34 Mattia Righi, <notifications@github.com> wrote:

> I'm a bit confused.
> I've tried to run:
>
> esmvaltool -cm -c config-user.yml DUMMY -ol DUMMY
>
> I understand the the first DUMMY is a workaround to replace the recipe,
> but I would expect the second DUMMY to run the script
> cmor/cmorize_obs/cmorize_obs_DUMMY.ncl.
> This is what is shown on stdout:
>
> 2018-10-19 07:30:11,789 UTC [36187] INFO    Starting the ESMValTool CMORization Mode v2.0a1 at time: 2018-10-19 07:30:11 UTC
> 2018-10-19 07:30:11,789 UTC [36187] INFO    ----------------------------------------------------------------------
> 2018-10-19 07:30:11,789 UTC [36187] INFO    INPUT DIR    = /ESMVal/obs/RAW
> 2018-10-19 07:30:11,790 UTC [36187] INFO    OUTPUT DIR = /ESMValTool/v2_output/CMORization_20181019_073011
> 2018-10-19 07:30:11,790 UTC [36187] INFO    ----------------------------------------------------------------------
> 2018-10-19 07:30:11,790 UTC [36187] INFO    Running the CMORization scripts.
> 2018-10-19 07:30:11,791 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
> 2018-10-19 07:30:11,792 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
> 2018-10-19 07:30:11,792 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
> 2018-10-19 07:30:11,793 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
> 2018-10-19 07:30:11,793 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
> 2018-10-19 07:30:11,793 UTC [36187] INFO    Ending the Earth System Model Evaluation Tool v2.0a1 at time: 2018-10-19 07:30:11 UTC
> 2018-10-19 07:30:11,793 UTC [36187] INFO    Time for running the recipe was: 0:00:00.004039
> 2018-10-19 07:30:11,794 UTC [36187] INFO    Finished the reformatting module
> 2018-10-19 07:30:11,794 UTC [36187] INFO    Run was succesful
>
> What am I doing wrong?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-431272365>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCo9Ov1HAP2EFV80frwY0ayDBByDM9ks5umYCRgaJpZM4XkAe9>
> .
>

valeriupredoi commented:
Sorry, in RAW, in your case, not RAWOBS

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 19 Oct 2018, 08:40 Valeriu Predoi, <valeriu.predoi@gmail.com> wrote:

> That's because you dont have any data directory called DUMMY in RAWOBS,
> Mattia. The -ol is the list of datasets that need to be cmorized but if
> there is no dataset with those names in RAWOBS, it will not run
>
> Dr Valeriu Predoi.
> Computational scientist
> NCAS-CMS
> University of Reading
> Department of Meteorology
> Reading RG6 6BB
> United Kingdom
>
> On Fri, 19 Oct 2018, 08:34 Mattia Righi, <notifications@github.com> wrote:
>
>> I'm a bit confused.
>> I've tried to run:
>>
>> esmvaltool -cm -c config-user.yml DUMMY -ol DUMMY
>>
>> I understand the the first DUMMY is a workaround to replace the recipe,
>> but I would expect the second DUMMY to run the script
>> cmor/cmorize_obs/cmorize_obs_DUMMY.ncl.
>> This is what is shown on stdout:
>>
>> 2018-10-19 07:30:11,789 UTC [36187] INFO    Starting the ESMValTool CMORization Mode v2.0a1 at time: 2018-10-19 07:30:11 UTC
>> 2018-10-19 07:30:11,789 UTC [36187] INFO    ----------------------------------------------------------------------
>> 2018-10-19 07:30:11,789 UTC [36187] INFO    INPUT DIR    = /ESMVal/obs/RAW
>> 2018-10-19 07:30:11,790 UTC [36187] INFO    OUTPUT DIR = /ESMValTool/v2_output/CMORization_20181019_073011
>> 2018-10-19 07:30:11,790 UTC [36187] INFO    ----------------------------------------------------------------------
>> 2018-10-19 07:30:11,790 UTC [36187] INFO    Running the CMORization scripts.
>> 2018-10-19 07:30:11,791 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
>> 2018-10-19 07:30:11,792 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
>> 2018-10-19 07:30:11,792 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
>> 2018-10-19 07:30:11,793 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
>> 2018-10-19 07:30:11,793 UTC [36187] INFO    No need to CMORize, could not find CMOR script.
>> 2018-10-19 07:30:11,793 UTC [36187] INFO    Ending the Earth System Model Evaluation Tool v2.0a1 at time: 2018-10-19 07:30:11 UTC
>> 2018-10-19 07:30:11,793 UTC [36187] INFO    Time for running the recipe was: 0:00:00.004039
>> 2018-10-19 07:30:11,794 UTC [36187] INFO    Finished the reformatting module
>> 2018-10-19 07:30:11,794 UTC [36187] INFO    Run was succesful
>>
>> What am I doing wrong?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-431272365>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AbpCo9Ov1HAP2EFV80frwY0ayDBByDM9ks5umYCRgaJpZM4XkAe9>
>> .
>>
>

mattiarighi commented:
The problem is that our RAWOBS directory is structured in subdirectories with Tiers (Tier1, Tier2, Tier3).
So in general I think the existence of the `[dataset]` subdirectory in RAWOBS should not be a precondition to run `cmorize_obs_[dataset]`.

valeriupredoi commented:
Ah yes, this is why you need to pass RAWOBS: /path/Tier1 or 2 or 3. The way
it is implemented now is it needs a match between -ol datasets, the
cmorization scripts of those datasets and the existence of the datasets
themselves in RAWOBS; this way we are sure we are doing the right thing
with minimum amount of code and no if/else conditions.

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 19 Oct 2018, 08:59 Mattia Righi, <notifications@github.com> wrote:

> The problem is that our RAWOBS directory is structured in subdirectories
> with Tiers (Tier1, Tier2, Tier3).
> So in general I think the existence of the [dataset] subdirectory in
> RAWOBS should not be a precondition to run cmorize_obs_[dataset].
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-431278750>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCo0WAEiF6XkQAf82ugaVADmgsFR4oks5umYZjgaJpZM4XkAe9>
> .
>

mattiarighi commented:
Ok, now I've added `Tier2/` to the path and created a `DUMMY/` subdir in it.
But I still don't manage to ge my script running...
```
esmvaltool -cm -c config-user_PA2.yml DUMMY -ol DUMMY

Starting the ESMValTool CMORization Mode v2.0a1 at time: 2018-10-19 08:36:22 UTC
----------------------------------------------------------------------
INPUT DIR    = /ESMVal/obs/RAW/Tier2
OUTPUT DIR = /ESMValTool/v2_output/CMORization_20181019_083622
----------------------------------------------------------------------
Running the CMORization scripts.
No need to CMORize, could not find CMOR script.
No need to CMORize, could not find CMOR script.
No need to CMORize, could not find CMOR script.
No need to CMORize, could not find CMOR script.
No need to CMORize, could not find CMOR script.
Ending the Earth System Model Evaluation Tool v2.0a1 at time: 2018-10-19 08:36:22 UTC
Time for running the recipe was: 0:00:00.002188
Finished the reformatting module
Run was succesful
```

valeriupredoi commented:
Haha, you apso need the data file, man, the thing is smart, won't run
unless there is a var_DUMMY_* data file there. Note that it will take care
of the variables if there's more than one

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 19 Oct 2018, 09:38 Mattia Righi, <notifications@github.com> wrote:

> Ok, now I've added Tier2/ to the path and created a DUMMY/ subdir in it.
> But I still don't manage to ge my script running...
>
> esmvaltool -cm -c config-user_PA2.yml DUMMY -ol DUMMY
>
> Starting the ESMValTool CMORization Mode v2.0a1 at time: 2018-10-19 08:36:22 UTC
> ----------------------------------------------------------------------
> INPUT DIR    = /ESMVal/obs/RAW/Tier2
> OUTPUT DIR = /ESMValTool/v2_output/CMORization_20181019_083622
> ----------------------------------------------------------------------
> Running the CMORization scripts.
> No need to CMORize, could not find CMOR script.
> No need to CMORize, could not find CMOR script.
> No need to CMORize, could not find CMOR script.
> No need to CMORize, could not find CMOR script.
> No need to CMORize, could not find CMOR script.
> Ending the Earth System Model Evaluation Tool v2.0a1 at time: 2018-10-19 08:36:22 UTC
> Time for running the recipe was: 0:00:00.002188
> Finished the reformatting module
> Run was succesful
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-431289178>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCo9qnHEFaYuI7KaBeBhyq2AZ_u0Ebks5umY-MgaJpZM4XkAe9>
> .
>

mattiarighi commented:
This is getting complicated: the reformat scripts from v1 already take care of all these details.
They just need to gather the information about the input path / output path from the interface and that's it.

There are even some special cases, where no input data are needed and the output is created applying fitting functions with some published parameters. These are given in the script itself (see, e.g., [this one](https://github.com/ESMValGroup/ESMValTool/blob/development/reformat_scripts/obs/reformat_obs_Putaud.ncl)).

We need to support such high flexibility, since the best way to cope with the huge zoo of obs data formats is to have dedicated stand-alone scripts for each dataset.

valeriupredoi commented:
I dont understand why it is complicated, all you need is to specify (if you
want to) -ol DATASETS_LIST and as long as they exist in your RAWOBS as data
files and the cmorization scripts exist, they will run, as simple as that.
Why would you run a cmorization script on a non-txistant data file? Am I
missing smth in here? If no -ol in command line then it will run all
cmorization scripts on all matching datasets

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 19 Oct 2018, 09:49 Mattia Righi, <notifications@github.com> wrote:

> This is getting complicated: the reformat scripts from v1 already take
> care of all these details.
> They just need to gather the information about the input path / output
> path from the interface and that's it.
>
> There are even some special cases, where no input data are needed and the
> output is created applying fitting functions with some published
> parameters. These are given in the script itself (see, e.g., this one
> <https://github.com/ESMValGroup/ESMValTool/blob/development/reformat_scripts/obs/reformat_obs_Putaud.ncl>
> ).
>
> We need to support such high flexibility, since the best way to cope with
> the huge zoo of obs data formats is to have dedicated stand-alone scripts
> for each dataset.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-431292192>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCoy5DrR0jajEGh4Bl-kYFUVr90wTMks5umZH_gaJpZM4XkAe9>
> .
>

mattiarighi commented:
>  dont understand why it is complicated, all you need is to specify (if you want to) -ol DATASETS_LIST and as long as they exist in your RAWOBS as data files and the cmorization scripts exist, they will run, as simple as that. 

I'm just saying that the existence of the input path / files should not be a pre-condition for running the scripts, since there are cases where the observational data are not converted/cmorized from an existing raw datasets, but rather generated using fitting functions from parameters published in a paper. I would simply set the existence of `reformat_obs_[dataset]` as a condition.

> Why would you run a cmorization script on a non-txistant data file? Am I missing smth in here? If no -ol in command line then it will run all cmorization scripts on all matching datasets

See the example I've posted above: the data used to generate these observations are not available in any digital form, they appear as a table in a published report. The only way to get a NetCDF file out of it was to *manually* copy the data into a script. 

I know it is weird, but for some dataset (especially some old ones) there is no other solution.
There are even cases of data in Excel format, and the user has to manually extract a CSV file from it before applying the script.

> Note that it will take care of the variables if there's more than one

This point is unclear to me. How is it taking care of the variables? Is it expecting CMOR variable names in the raw filenames?

Hope this helps to clarify.

valeriupredoi commented:
ok @mattiarighi thanks for illuminating me on the phone, will make the changes and post a synopsis in a jiffy, this OBS business is as fussy as a hen with one chick

valeriupredoi commented:
oh and for beauty purposes we should run `esmvaltool -cr -c config-user.yml CMOR -ol DAT1,DAT2...` since that argument can be called anything, and if we call it CMOR it's obvious for the user esmvaltool gets into CMORization mode, and it looks fancy and smart :grin:

mattiarighi commented:
Thanks @valeriupredoi it works now! :tada: 
I've made some very minor changes to the logger messages.

It would be nice to get rid of the dummy argument replacing the recipe argument, if possible.
And the help message should also be updated accordingly.

I leave this to @bouweandela for further comments.
I will start porting some actual scripts in a separate branch once this is merged.

valeriupredoi commented:
@mattiarighi awesome!! I can drink my Friday beers in peace :grin: -- so about the dummy recipe argument, I have actually spent a good hour fiddling with `parser` trying to make it not ask for the recipe AND still keep the possibility to pass it without a shortop/longop flag - the problem is that once you call `parser.parse_args()` to get the arguments in the form of a list, it's game over and can't change the parser anymore even if you reassign it or copy it. I honestly don't know how to make both cases work, unless we give `recipe` a short/long argument like it used to have `-r/--recipe`, in that case it works fine for both cases and no need to specify the required recipe when -n `-cm` mode (note that for `-r/--recipe` we should make `required=False` or not write the required argument at all). The check for recipe requirement can then be made inside the recipe/diag mode in _main

bouweandela commented on esmvaltool/reformat.py:
```diff
@@ -0,0 +1,139 @@
+                logger.info("CMORizing dataset %s using NCL script %s",
+                            dataset, reformat_script)
+                # copy over the reformat script
+                subprocess.call(['cp', reformat_script, out_data_dir])
```
use the `shutil` module instead of the shell `cp` command

bouweandela commented:
See this https://github.com/ESMValGroup/ESMValTool/issues/232#issuecomment-406546739 comment for a suitable name for the executable. If you would really like to start the command with `esmvaltool`, you can also follow this suggestion https://github.com/ESMValGroup/ESMValTool/issues/232#issuecomment-406574796.

valeriupredoi commented:
@bouweandela cheers, dude, I've started working on implementing your comments, but since they are logistical and not influencing the functionality as is right now -- @mattiarighi I'd suggest you fire up the reformatting with this current workflow for the sake of gaining a day or two (if it's urgent)

mattiarighi commented:
I won't be able to start until next week, so let's address @bouweandela 's comments before merging.

valeriupredoi commented:
yooo guys @mattiarighi and @bouweandela -- done, we have us a nice executable now that I tested and the behaviour is the same as the previous implementation, cheers, Bouwe for the suggestions! Mattia, run: `cmorize_observations --help` to see the options, it's pretty easy - `cmorize_observations -c config-user.yml` to run on ALL the obss in RAW_OBS or `cmorize_observations -c config-user.yml -o OBS1,OBS2...` to run on a select list of obss

mattiarighi commented:
Should I try to add a real script to your branch as a test case?

valeriupredoi commented:
si si, try it @mattiarighi !

mattiarighi commented:
I've added the script for AURA-TES data as a test case.
One remark: the interface information are saved in  `output_dir_path/<dataset>_info.ncl`.
It would be good to define this file as an environment  variable in `settings.ncl` as it is done for the diagnostics. This way, you can just load the file using always the same line, independently of the dataset in use.

valeriupredoi commented:
cool @mattiarighi I shall look into this right now

valeriupredoi commented:
hey @mattiarighi you now get a `CMOR/CMOR_[DATE]_[TIME]/run/settings.ncl` written when the executable runs, it contains three NCL variables: `cmorization_script, input_dir_path, output_dir_path` - is that what you need? Note that `settings.ncl` gets recreated with every new dataset reformatting pass

mattiarighi commented:
- could you put this in `CMOR/CMOR_[DATE]_[TIME]/run/[DATASET]/settings.ncl`, to avoid overwriting?
- If we want to use the logging function we also need to pass `config_user_info@log_level`, but I'm not sure how the logging messages are handled here.

valeriupredoi commented:
done the first point, have absolutely no clues how the `logger.logging` module is handles in NCL, you the boss of that or @bouweandela

mattiarighi commented:
It collects all messages from the NCL execution and saves them in a logfile in the run_dir.
The question is whether we want to use such approach also for the cmorization scripts.

valeriupredoi commented:
ok @mattiarighi you now have logging to file as well as to screen, use it wisely :rofl:

mattiarighi commented:
I'm still not able to load the `settings.ncl` file within the script. I think we need something like this:
https://github.com/ESMValGroup/ESMValTool/blob/b52ebd0437f6e614f59fb7e1b27a2f864b7f0874/esmvaltool/_task.py#L420-L423
in `cmorize_observations.py`.

Then I can load them in NCL using
```
loadscript("$settings")
````
Does it make sense?

valeriupredoi commented:
ow yes, you are right, it's not an environment variable, it's just a file; yeah, if you add that code it will be an env variable, hang on will do iot now. Sorry, forgot that should be an env variable :stuck_out_tongue:

valeriupredoi commented:
done and tested!

mattiarighi commented:
Alright, now it finds the settings file.
But it fails to locate the utility file in the same directory:
```
load "./cmor/cmorize_obs/cmorize_obs_utils.ncl"
```
Are we missing some path?

valeriupredoi commented:
careful where you loading from - remember the execution is done inside the $user/CMOR/CMOR_$date_$time/TierX/DATASET - if you need to load NCL libraries we need to make those env variesbles or place them inside the dir where the execution takes places

mattiarighi commented:
Is there a way to set `esmvaltool/` as the root path, as we did for the diagnostics?
In that case you could load anything just giving the relative path to `esmvaltool/`, e.g.:
https://github.com/ESMValGroup/ESMValTool/blob/b52ebd0437f6e614f59fb7e1b27a2f864b7f0874/esmvaltool/diag_scripts/perfmetrics/main.ncl#L85-L91

valeriupredoi commented:
you have now an `esmvaltool_root` environment variable that points to the esmvaltool installed root directory, so you can do `loadscript("$esmvaltool_root/cmor/cmorize_obs/cmorize_obs_utils.ncl")`

mattiarighi commented:
Ok. now I'm able to load the scripts I need.

There is however a problem with the `logging.ncl`, which is trying to load other interface scripts which are not available from this path:

https://github.com/ESMValGroup/ESMValTool/blob/b52ebd0437f6e614f59fb7e1b27a2f864b7f0874/esmvaltool/interface_scripts/logging.ncl#L23

This piece of code is outdated (it will be replaced by provencance tracking #649).
I will open a PR to remove this part and then continue here.

valeriupredoi commented:
Cool, is there anything I need to do about this? Can you run the
cmorization ok? Let's get those babies rolling :grin:

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Tue, 30 Oct 2018, 08:59 Mattia Righi <notifications@github.com wrote:

> Ok. now I'm able to load the scripts I need.
>
> There is however a problem with the logging.ncl, which is trying to load
> other interface scripts which are not available from this path:
>
>
> https://github.com/ESMValGroup/ESMValTool/blob/b52ebd0437f6e614f59fb7e1b27a2f864b7f0874/esmvaltool/interface_scripts/logging.ncl#L23
>
> This piece of code is outdated (it will be replaced by provencance
> tracking #649 <https://github.com/ESMValGroup/ESMValTool/pull/649>).
> I will open a PR to remove this part and then continue here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-434220954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCowRoBVDNrahsR2DBToF6nN-r-h1xks5uqBT3gaJpZM4XkAe9>
> .
>

mattiarighi commented:
I cannot test the logging with this problem.
Let me tackle it first and then get back here.

mattiarighi commented:
Alright, PR #684 opened.
Waiting for approval and then will continue here.
Stay tuned :radio:

valeriupredoi commented:
:radio_button: tuning in :grin:

valeriupredoi commented:
@mattiarighi back from the UKESM retreat, que pasa in here?

mattiarighi commented:
I'm busy with #692 which is urgently needed by other users and then will get back here.

valeriupredoi commented:
take your time @mattiarighi lemme know if I can help

valeriupredoi commented:
:tumbleweeds: :ghost: this thread looks totally deserted :grin:

mattiarighi commented:
Finally getting back to this... :hourglass_flowing_sand: 

I've start adapting the script for the AURA-TES data, as a working example.
I've put the raw data in the RAWOBS directory (under Tier2) as defined in my `config-user.yml` file.
The scripts is found and executed, but there is a problem with the interface file (`cmor/cmorize_obs/interface.ncl`), which fails to load the logging file. 

Apparently some interface variables are not passed correctly:

```
2018-12-10 09:23:40,528 INFO     esmvaltool.utils.cmorize_observations,120      Executing cmd: ncl cmorize_obs_AURA-TES.ncl
2018-12-10 09:23:44,645 INFO     esmvaltool.utils.cmorize_observations,125      [NCL] b' Copyright (C) 1995-2018 - All Rights Reserved
2018-12-10 09:23:44,646 INFO     esmvaltool.utils.cmorize_observations,125      [NCL]  University Corporation for Atmospheric Research
2018-12-10 09:23:44,647 INFO     esmvaltool.utils.cmorize_observations,125      [NCL]  NCAR Command Language Version 6.5.0
2018-12-10 09:23:44,647 INFO     esmvaltool.utils.cmorize_observations,125      [NCL]  The use of this software is governed by a License Agreement.
2018-12-10 09:23:44,647 INFO     esmvaltool.utils.cmorize_observations,125      [NCL]  See http://www.ncl.ucar.edu/ for more details.
2018-12-10 09:23:44,648 INFO     esmvaltool.utils.cmorize_observations,125      [NCL] fatal:Variable (config_user_info) is still undefined, unable to reference attribute log_level
2018-12-10 09:23:44,648 INFO     esmvaltool.utils.cmorize_observations,125      [NCL] \rfatal:["Execute.c":8637]:Execute: Error occurred at or near line 77 in file /esmvaltool/utils/../interface_scripts/logging.ncl
```

valeriupredoi commented:
@mattiarighi good to see this being taken out the mothballs :grin: - is the error related to the python executable or to the NCL reformatting script that is executed; I'll attempt at replicating the error and will have a closer look now

valeriupredoi commented:
@mattiarighi I just added the `config_user_ino@log_level` ncl variable and tested with AURA-TES and it goes through without an error. Sorries, man, I was not aware this variable needs to be initialized, me no speaky ncl :grin:

mattiarighi commented:
Another minor suggestions: can we rename the path `cmor/cmorize_obs/` to `cmor/cmorizers/obs/` ?
This is because I expect additional cmorizers to be added in the future (e.g., the model specific ones, see #233).

You could also think about putting your freshly-merged `mip_converter` in this directory, rather than in `utils/`.

valeriupredoi commented:
hey @mattiarighi good point! can you go ahead and make the change? I wouldn't put the `mip_convert` tool there because it's actually a cross-package situation that may or may not work always (depends if the actual `mip_convert` is installed on site) so we don't want to leave the impression that some cmorizers may not work. `utils` is a relatively low level directory, so I'd keep it there for now

valeriupredoi commented:
> Finally getting back to this, sorry for the long delay.
> 
> I've ported the cmorizing script for `AURA-TES`: it still needs a bit of polishing and fixing some internal errors, but it runs, the error messages and the logging now work as expected.
> Only the log output looks a bit odd, as it contains special characters (like `\t`) and the whole NCL logging is enclosed in a block `b'...'` (why `b`, just a typo?). Here an example:
> 
> ```
> Executing cmd: ncl cmorize_obs_AURA-TES.ncl
> 17:15:10 INFO     esmvaltool.utils.cmorize_observations,128     [NCL] b' Copyright (C) 1995-2018 - All Rights Reserved
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL]  Copyright (C) 1995-2018 - All Rights Reserved
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL]  University Corporation for Atmospheric Research
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL]  NCAR Command Language Version 6.5.0
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL]  The use of this software is governed by a License Agreement.
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL]  See http://www.ncl.ucar.edu/ for more details.
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL] (0)\tINFO    Processing date 200501
> ...
> ...
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL] (0)\tINFO    Processing date 201112
> 17:15:11 INFO     esmvaltool.utils.cmorize_observations,128     [NCL] '
> ```
> 
> Another minor suggestion: can we rename the output directory to `cmorize_observations_YYYYMMDD_HHMMSS`? This would be more consistent I think.

This is awesome!! Cool stuff! The `b...` because of bytesize object rather than string, but it's the same way it's handled.

mattiarighi commented:
> hey @mattiarighi good point! can you go ahead and make the change?

Done. I've also renamed the executable to `cmorize_obs`, more compact and consistent with the directory names in `cmor/cmorizers/`. In the future we can add `cmorize_EMAC` and other model-specific cmorizers.

> I wouldn't put the mip_convert tool there because it's actually a cross-package situation that may or may not work always (depends if the actual mip_convert is installed on site) so we don't want to leave the impression that some cmorizers may not work. utils is a relatively low level directory, so I'd keep it there for now

Ok.

> The b... because of bytesize object rather than string, but it's the same way it's handled.

Can you change it and remove the special characters `\t` etc. ?

Another thing which is not clear to me: why is the LOGLEVEL passed as optional argument. Shouldn't it be read from the config-user file?

Otherwise I have no further comments and I leave it to @bouweandela for review. Thanks!

valeriupredoi commented:
@mattiarighi you are absolutely right: LOG_LEVEL is now extracted from the config file, this was a legacy command line argument (nice way of saying I completely forgot to remove it). Unfortunately I got no idea how to get rid of the b`, \t and \r characters, they come from NCL and I don't know how to tell the Python logger to ignore them. @bouweandela do you know?

bouweandela commented:
> Unfortunately I got no idea how to get rid of the b`, \t and \r characters,

Use `universal_newlines=True` as a keyword argument with `subprocess.Popen`.

bouweandela commented on esmvaltool/utils/cmorize_obs.py:
```diff
@@ -0,0 +1,273 @@
+                logger.info("CMORizing dataset %s using Python script %s",
+                            dataset, py_reformat_script)
+                # copy over the reformat script
+                shutil.copy2(py_reformat_script,
```
Why are you copying and then importing? This may lead non-reproducible results if people start editing those files and re-running. Best to just import from the installed esmvaltool.

valeriupredoi replied:
same as with the ncl script, copying is fine coz it gets overwritten with the installed script everytime one runs :)

valeriupredoi replied:
le fixed

bouweandela commented on esmvaltool/utils/cmorize_obs.py:
```diff
@@ -0,0 +1,273 @@
+                logger.info("CMORizing dataset %s using NCL script %s",
+                            dataset, reformat_script)
+                # copy over the reformat script
+                shutil.copy2(reformat_script, out_data_dir)
```
Why are you copying and then running? This may lead non-reproducible results if people start editing those files and re-running. Best to just run from the installed esmvaltool.

bouweandela replied:
Just provide an absolute path to the script in the `Popen` command and it should work. If you need to load additional NCL scripts from the main NCL script, you can store the location of the reformat script directory in an environmental variable, e.g. reformat_scripts=/some/path/esmvaltool/cmor/cmorize and use `load "$reformat_scripts/obs/some_reformat_script.ncl"` in NCL for loading.

valeriupredoi replied:
le fixed

bouweandela commented on esmvaltool/utils/cmorize_obs.py:
```diff
@@ -0,0 +1,273 @@
+    logger.info("Processing datasets %s", datasets)
+
+    # loop through tier/datasets to be cmorized
+    for tier, _ in datasets.items():
```
```suggestion
    for tier in datasets:
```

bouweandela commented on esmvaltool/utils/cmorize_obs.py:
```diff
@@ -0,0 +1,273 @@
+The CMOR reformatting scripts are to be found in:
+esmvaltool/cmor/cmorizers/obs
+
+    Usage
```
Manually typing the Usage section is a bad idea, as it will get out of sync with the actual options available. If you just provide the `help='Some explanation'` argument to the `parser.add_argument` method, `argparse` will create this section for you when you run `mytool --help`.

valeriupredoi commented:
> > Unfortunately I got no idea how to get rid of the b`, \t and \r characters,
> 
> Use `universal_newlines=True` as a keyword argument with `subprocess.Popen`.

aha! but that removes the logging formatter (date, time, line, tab) from the NCL output and prints only the NCL output lines...

bouweandela commented:
Maybe see: https://stackoverflow.com/questions/606191/convert-bytes-to-a-string

valeriupredoi commented:
> Maybe see: https://stackoverflow.com/questions/606191/convert-bytes-to-a-string

yup, that done it nicely, man, cheers :beer:

valeriupredoi commented on esmvaltool/utils/cmorize_obs.py:
```diff
@@ -0,0 +1,273 @@
+    logger.info("Executing cmd: %s", ' '.join(ncl_call))
+    process = subprocess.Popen(ncl_call, stdout=subprocess.PIPE,
+                               stderr=subprocess.STDOUT, env=env)
+    output, _ = process.communicate()
```
captured and outputted

valeriupredoi commented:
@mattiarighi you now have clean and pretty logging output without special characters (I liked those :grin: ) thanks to Bouwe's suggestion :)

bouweandela commented:
I would prefer to have all cmorization code used by this tool in one place, e.g. in a directory esmvaltool/utils/cmorize, instead of spread out over esmvaltool/utils/cmorize.py and esmvaltool/cmor/cmorizers.

mattiarighi commented:
> I would prefer to have all cmorization code used by this tool in one place, e.g. in a directory esmvaltool/utils/cmorize, instead of spread out over esmvaltool/utils/cmorize.py and esmvaltool/cmor/cmorizers.

Then let's put it in `esmvaltool/utils/cmorizers/obs/` and also add the mip-converter stuff there (`cmorizers/mip-converter/` or similar). As suggested above, it would be nice to have all cmorizers in one place, also in view of #233.

mattiarighi commented:
@valeriupredoi can you address the [two](https://github.com/ESMValGroup/ESMValTool/pull/666#pullrequestreview-190290991) [remaining](https://github.com/ESMValGroup/ESMValTool/pull/666#pullrequestreview-190291295) issues and move the files as suggested [here](https://github.com/ESMValGroup/ESMValTool/pull/666#issuecomment-452600474)?

valeriupredoi commented:
@mattiarighi on it now

valeriupredoi commented:
OK, done! @mattiarighi @bouweandela :eyes:

mattiarighi commented:
I still see the files in `cmor/cmorizeres/obs/`...

valeriupredoi commented:
@mattiarighi didnt move those - I reckon those should stay there, they are not tools/utilities, you guys want all those moved? I think this is a good structure as is now

mattiarighi commented:
I think it would be better to have all files in cmorizers/obs

valeriupredoi commented:
they reluctantly moved there :grin: and I tested, it seems to work fine :tomato:

mattiarighi commented:
Thanks @valeriupredoi! :beer: 

I've to slightly adapt `cmorize_obs_ESACCI-LANDCOVER.ncl` as requested by @tobstac in #756 and then will merge this (if @bouweandela does not have further remarks).

valeriupredoi commented:
awesome! if @bouweandela still has remarks he will not get :beer: at DLR from me :grinning:

mattiarighi commented:
One very last thing: I've noticed that an empty (`None`) error message is output at the end of the NCL scripts.
Any idea where does it come from?

```
2019-01-15 15:51:43,813 INFO     esmvaltool.utils.cmorizers.obs.cmorize_obs,116 [NCL]
2019-01-15 15:51:43,813 INFO     esmvaltool.utils.cmorizers.obs.cmorize_obs,117 [NCL][ERROR] None
2019-01-15 15:51:43,814 INFO     esmvaltool.utils.cmorizers.obs.cmorize_obs,203 Ending the CMORization Tool at time: 2019-01-15 14:51:43 UTC
2019-01-15 15:51:43,814 INFO     esmvaltool.utils.cmorizers.obs.cmorize_obs,206 Time for running the CMORization scripts was: 0:00:31.319920
```

valeriupredoi commented:
yes, that's the `err` bit of the call to `Popen`, basically telling us that all went well with the subprocess call from python to NCL

valeriupredoi commented:
@mattiarighi quick question: I am trying to use the workflow to cmorize `/group_workspaces/jasmin4/esmeval/obsdata/Tier2/AURA-TES/OBS_AURA-TES_sat_1_T3M_tro3_200501-200912.nc` which is on jasmin, and belongs to the DKRZ mirror that @bjoernbroetz built some time ago - it's not working saying `fatal:Variable (outvar) is undefined` and also spits out warnings at every month between 2005 and 2011(!?) saying file not found

mattiarighi commented:
You need to download the new rawobs from the source, I updated the script to work with the current version of the data. Just follow the instructions in the script header.
Some months are skipped because data are not available and the field is filled with missing, but they are just a few.

valeriupredoi commented:
cool, if it works well for you that's important, I really don't need any AURA-TES files :grin: we should make sure the Jasmin mirror is up to date when all this CMORization business has been completed - @bjoernbroetz

mattiarighi commented:
Definitely.

bjoernbroetz commented:
> ... and belongs to the DKRZ mirror that @bjoernbroetz built some time ago - it's not working ...

There are no raw data on the Jasmin repo that I made. Just the reformatted ones...

valeriupredoi commented:
@bjoernbroetz yeah sorry, man, talked out of my arse in that comment, what I meant was let's make sure the newly cmorized OBS's that @mattiarighi is producing right now will get mirrored on to Jasmin as well

valeriupredoi commented:
...as in, I was trying to cmorize an already cmorized file, go me :shipit:

bouweandela commented on esmvaltool/utils/cmorizers/obs/cmorize_obs.py:
```diff
@@ -0,0 +1,262 @@
+    output, err = process.communicate()
+    for oline in str(output.decode('utf-8')).split('\n'):
+        logger.info('[NCL] %s', oline)
+    logger.info('[NCL][ERROR] %s', err)
```
```suggestion
    if err:
        logger.warning('[NCL][ERROR] %s', err)
```

bouweandela commented on esmvaltool/utils/cmorizers/obs/cmorize_obs.py:
```diff
@@ -0,0 +1,262 @@
+                py_reformat_script = reformat_script_root + '.py'
+                logger.info("CMORizing dataset %s using Python script %s",
+                            dataset, py_reformat_script)
+                sys.path.append(reformat_scripts)
```
Use [`importlib.import_module`](https://docs.python.org/3/library/importlib.html#importlib.import_module) instead of messing with `sys.path`.

bouweandela commented on esmvaltool/utils/cmorizers/obs/cmorize_obs.py:
```diff
@@ -0,0 +1,262 @@
+    datasets = {}
+
+    # get all available tiers in source dir
+    for _, tier, _ in os.walk(raw_obs, followlinks=True):
```
It would be better to be a bit more restrictive here, e.g.
```python
tiers = ['Tier{}'.format(i) for i in range(1, 4)]
tiers = [tier for tier in tiers if os.path.exists(os.path.join(raw_obs, tier))]
```

mattiarighi replied:
But note that `Tier` ranges between 1 and 3.
Actually 2 and 3, since `Tier1` is `obs4mips` which is not supposed to be cmorized.

bouweandela commented on esmvaltool/utils/cmorizers/obs/cmorize_obs.py:
```diff
@@ -0,0 +1,262 @@
+
+    # if user specified obs list
+    if obs_list:
+        for tier in tiers[0]:
```
Why are you only using the first tier found? Or is the name `tiers` misleading?

bouweandela commented on esmvaltool/utils/cmorizers/obs/cmorize_obs.py:
```diff
@@ -0,0 +1,262 @@
+    else:
+        for tier in tiers[0]:
+            datasets[tier] = []
+            for _, dats, _ in os.walk(os.path.join(raw_obs, tier),
```
This will not work correctly with unexpected directory structures. Use `os.listdir(os.path.join(raw_obs, tier)` to list the available datasets instead.

valeriupredoi commented:
very good calls @bouweandela - cheers, man, plugged them all and tested thoroughly. @mattiarighi looping only on `[Tier2, Tier3]` as you pointed out there, didn't know Tier1 doesn't need craporization :grin:

bouweandela commented:
Code looks fine now, but I did not test if it works.

valeriupredoi commented:
cheers @bouweandela - it works well, I tested quite a bit and @mattiarighi has already run the cmorization for 2 obs models; Mattia, let me know if you need help with more cmorization scripts (eg the Python ones, not NCL pls :grin: )
